wgsl/index.bs

```diff
@@ -4557,11 +4555,291 @@ TODO: *Stub*: Expression evaluation
+* If "Value_return" exists, look at which nodes are reachable from it
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+
+Note: Anything which is not explicitly remembered, including the entire graph can be destroyed at this point. The little information that we do need to remember to analyze the caller is linear in the number of parameters of the function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
+
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+We achieve this by assigning to each statement a set of *behaviors* out of {Return, Break, Continue, Discard, Fallthrough, Nothing}.
+Intuitively, this set corresponds to the ways that control flow can exit the statement.
 
-### Divergence and reconvergence TODO ### {#divergence-reconvergence}
+TODO: do we need an extra behavior for possibly infinite loops? I should check whether some form of reconvergence happens after something like `while(tid mod 2) {}`.
 
-### Uniformity restrictions TODO ### {#uniformity-restrictions}
+If the set of behaviors for a statement has a single element, this means that we either did not diverge within the statement, or we reconverged, and we pick the node corresponding to control flow at the start of the statement as the node corresponding to control flow at the exit of the statement, overriding the other rules below.
+
+The rules for analyzing statements take as argument both the statement itself and the node corresponding to control flow at the beginning of it (which we'll note "CF" below) and return the following:
+
+* A set of behaviors for the statement
+* A node corresponding to control flow at the exit of it
+* A set of new nodes and edges to add to the graph
+
+In the table below, `CF1, S -> CF2, B` means "run the analysis on S starting with control flow CF1, apply the required changes to the graph, and name the resulting control flow CF2, and the resulting set of behaviors B".
+Similarly, `CF1, E -> CF2, V, B` means "run the analysis on expression E, starting with control flow CF1, apply the required changes to the graph, and name the resulting control flow node CF2, the resulting value node V, and the resulting set of behaviors B" (see next section for the analysis of expressions).
+
+We have a similar set of rules for expressions in left-value positions, that we denote by `CF, E => CF, L, B`. Instead of computing the node which corresponds to the uniformity of the value, it computes the node which corresponds to the uniformity of the variable we are adressing.
+
+<table class='data'>
+  <caption>Uniformity rules for statements</caption>
+  <thead>
+    <tr><th>Statement<th>New nodes<th>Recursive analyses<th>Behaviors<th>Resulting control flow node<th>New edges
+  </thead>
+  <tr><td class="nowrap">*s1*; *s2*
+      <td>
+      <td class = "nowrap">*CF*, *s1* -> *CF1*, *B1*<br>
```

Comment:
Can we use parens for these tuples? `(*CF*, *s1*) -> (*CF1*, *B1*)`?
I was originally parsing this as `*CF*, (*s1* -> *CF1*), *B1*`

Reply:
Fixed, thanks for the suggestion.